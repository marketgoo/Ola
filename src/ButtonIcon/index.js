import React from 'react'
import {default as PT} from 'prop-types'
import { getElementType } from '../utils'
import cx from 'classnames'
import Icon from '../Icon'
import Spinner from '../Spinner'

const ButtonIcon = ({ as, className, busy, disabled, color, variant, invert, icon, ...props }) => {
  const ElementType = getElementType(ButtonIcon, { as: as, ...props })
  delete props.as

  const variantNumber = variant === 'normal' ? '500' : '900'

  const iconElement = typeof icon === 'string' ? <Icon name={icon} /> : icon

  const styles = cx(
    'ola_button_icon',
    `is-${variant}`,
    {'is-busy': busy},
    {'is-disabled': disabled && !busy},
    {'is-invert': invert},
    className
  )
  return (
    <ElementType
      className={styles}  
      disabled={busy || disabled}
      style={{
        '--base-color': `var(--color-${color}-${variantNumber})`,
        '--busy-color': `var(--color-${color}-600)`,
        '--disabled-color': `var(--color-${color}-300)`,
        '--shadow-focus': `var(--shadow-focus-${color})`,
      }}
      {...props}>
      {busy ? <Spinner /> : iconElement}
    </ElementType>
  )
}

ButtonIcon.defaultProps = {
  as: 'button',
  variant: 'normal',
  color: 'neutral',
  busy: false,
  disabled: false,
  invert: false,
  icon: null
}

ButtonIcon.propTypes = {
  /** Name of the icon in this lib (see Ola Icon), or a svg direcly **(preferred a Phosphor Icon)** */
  icon: PT.oneOfType([
    PT.string,
    PT.shape({
      type: PT.oneOf(['svg']),
    })
  ]),
  /** Render ButtonIcon with any html tag */
  as: PT.string,
  /** Icon color */
  color: PT.oneOf(['primary', 'neutral', 'positive', 'negative', 'warning', 'pro', 'accent']),
  /** Icon variant (normal=color-500, dark=color-900) */
  variant: PT.oneOf(['normal', 'dark']),
  /** Extra className */
  className: PT.string,
  /** Button busy state */
  busy: PT.bool,
  /** Button disabled state */
  disabled: PT.bool,
  /** Indicate if icon should have color white */
  invert: PT.bool,
}

export default ButtonIcon
