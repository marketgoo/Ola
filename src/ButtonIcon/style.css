.ola_buttonIcon {
    --color: var(--color-neutral-700);
    --background: transparent;
    --color-hover: var(--color-neutral-900);
    --background-hover: var(--color-neutral-200);
    --shadow-focus: var(--color-accent-100);
    --background-active: var(--color-neutral-400);

    display: inline-flex;
    min-height: var(--size-3);
    min-width: var(--size-3);
    align-items: center;
    justify-content: center;
    background: none;
    color: var(--color);
    border: none;
    cursor: pointer;
    padding: 0;
    outline: 0;
    vertical-align: middle;

    &::-moz-focus-inner {
        border: 0;
    }

    & > svg {
        display: block;
        flex-shrink: 0;
        max-height: 100%;
        max-width: 100%;
        border-radius: 50%;
        fill: var(--color);
        background-color: var(--background);
        transition: background-color var(--duration);
    }

    &:focus-visible svg {
        fill: var(--color-hover);
        box-shadow: 0 0 0 4px var(--shadow-focus);
    }

    &:hover svg {
        fill: var(--color-hover);
        background: var(--background-hover);
    }

    &:active svg {
        background: var(--background-active);
    }

    & .ola_icon.is-small {
        padding: var(--size-0);
    }
    & .ola_icon.is-medium,
    & .ola_spinner {
        padding: var(--size-1);
    }

    &.is-secondary {
        --color: var(--color-neutral-400);
    }

    &.is-destructive {
        --color: var(--color-negative-500);
        --shadow-focus: var(--color-negative-100);
        --color-hover: var(--color-negative-600);
        --background-active: var(--color-negative-100);
    }

    &.is-chevron {
        background-repeat: no-repeat;
        background-position: right center;
        padding-right: var(--size-4);
        border-radius: var(--radius-xl);

        /* Chevron gray */
        background-image: url('data:image/svg+xml;charset=UTF-8,<svg width="44px" height="44px" viewBox="0 0 44 44" version="1.1" xmlns="http://www.w3.org/2000/svg"><path d="M22,23.5857864 L27.2928932,18.2928932 C27.6834175,17.9023689 28.3165825,17.9023689 28.7071068,18.2928932 C29.0976311,18.6834175 29.0976311,19.3165825 28.7071068,19.7071068 L22.7071068,25.7071068 C22.3165825,26.0976311 21.6834175,26.0976311 21.2928932,25.7071068 L15.2928932,19.7071068 C14.9023689,19.3165825 14.9023689,18.6834175 15.2928932,18.2928932 C15.6834175,17.9023689 16.3165825,17.9023689 16.7071068,18.2928932 L22,23.5857864 Z" fill="%2357657a"></path></svg>');
    }

    &.is-disabled,
    &:disabled {
        color: var(--color-neutral-200);
        --background: var(--color-white);
        --background-hover: var(--color-white);
        --background-active: var(--color-white);
        cursor: not-allowed;
        pointer-events: none;
        box-shadow: none;

        &.is-chevron {
            /* Chevron gray-xlight */
            background-image: url('data:image/svg+xml;charset=UTF-8,<svg width="44px" height="44px" viewBox="0 0 44 44" version="1.1" xmlns="http://www.w3.org/2000/svg"><path d="M22,23.5857864 L27.2928932,18.2928932 C27.6834175,17.9023689 28.3165825,17.9023689 28.7071068,18.2928932 C29.0976311,18.6834175 29.0976311,19.3165825 28.7071068,19.7071068 L22.7071068,25.7071068 C22.3165825,26.0976311 21.6834175,26.0976311 21.2928932,25.7071068 L15.2928932,19.7071068 C14.9023689,19.3165825 14.9023689,18.6834175 15.2928932,18.2928932 C15.6834175,17.9023689 16.3165825,17.9023689 16.7071068,18.2928932 L22,23.5857864 Z" fill="%23f0f2f4"></path></svg>');
        }
    }

    &.is-busy {
        color: var(--color-white);
        background-color: var(--color-neutral-700);
        min-width: 0;
        cursor: wait;
        pointer-events: none;

        &.is-chevron {
            background-image: none;
        }
    }
}

