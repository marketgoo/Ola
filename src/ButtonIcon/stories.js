import React from 'react'
import ButtonIcon from './'
import { Copy, ShieldCheck, Airplane } from '@phosphor-icons/react'

export default {
  title: 'ButtonIcon',
  component: ButtonIcon,
  args: {
    icon: 'home',
    color: 'neutral',
    as: 'button',
    variant: 'regular',
    invert: false,
    disabled: false,
    busy: false,
  },
  argTypes: {
    onClick: { action: 'clicked' },
    onFocus: { action: 'focused' },
    color: {
      control: {
        type: 'select',
        options: ['primary', 'neutral', 'positive', 'negative', 'warning', 'pro', 'accent'],
      },
    },
    icon: {
      control: {
        type: 'select',
        options: ['add', 'analytics', 'arrowDown', 'arrowRight', 'arrowUp', 'chevronDown', 'chevronLeft', 'chevronRight', 'chevronUp', 'clock', 'close', 'copy', 'error', 'edit', 'help', 'home', 'informationCircle', 'link', 'newWindow', 'null', 'page', 'plus', 'post', 'search', 'site', 'staticPage', 'success', 'user', 'warning'], 
      },
    },
  },
  decorators: [
    (Story) => (
      <div style={{ backgroundColor: '#f6f6ff', padding: '32px' }}>
        {Story()}
      </div>
    ),
  ],
}

export const Base = (args) => <ButtonIcon {...args} />

export const PositiveColor = (args) => <ButtonIcon {...args} />
PositiveColor.args = {
  color: 'positive'
}

export const ProColor = (args) => <ButtonIcon {...args} />
ProColor.args = {
  color: 'pro'
}

export const Disabled = (args) => <ButtonIcon {...args} />
Disabled.args = {
  disabled: true
}

export const Busy = (args) => <ButtonIcon {...args} />
Busy.args = {
  busy: true
}

export const Invert = (args) => <ButtonIcon {...args} />
Invert.args = {
  invert: true
}

export const VariantDark = (args) => <ButtonIcon {...args} />
VariantDark.args = {
  variant: 'dark'
}

export const WithPhosphorIcons = (args) => <>
  <ButtonIcon {...args} icon={<Copy />} color="primary" />
  <ButtonIcon {...args} icon={<ShieldCheck />} color="positive" />
  <ButtonIcon {...args} icon={<Airplane />} color="pro" />
</>

