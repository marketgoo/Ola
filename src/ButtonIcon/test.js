import React from 'react'
import ButtonIcon from './'
import renderer from 'react-test-renderer'

it('Default ButtonIcon', () => {
  const tree = renderer
    .create(
      <ButtonIcon icon="success" color="positive" />
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
