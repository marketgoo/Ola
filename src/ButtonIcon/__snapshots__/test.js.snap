// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default ButtonIcon 1`] = `
<button
  className="ola_button_icon is-normal"
  disabled={false}
  style={
    Object {
      "--base-color": "var(--color-positive-500)",
      "--busy-color": "var(--color-positive-600)",
      "--disabled-color": "var(--color-positive-300)",
      "--shadow-focus": "var(--shadow-focus-positive)",
    }
  }
>
  <svg
    className="ola_icon is-medium"
    fillRule="evenodd"
    height="28"
    viewBox="0 0 28 28"
    width="28"
  >
    <path
      d="M14 2c1.63 0 3.186.317 4.666.95 1.48.634 2.756 1.486 3.827 2.557 1.07 1.071 1.923 2.347 2.557 3.827.633 1.48.95 3.035.95 4.666 0 1.63-.317 3.186-.95 4.666-.634 1.48-1.486 2.756-2.557 3.827-1.071 1.07-2.347 1.923-3.827 2.557-1.48.633-3.035.95-4.666.95-1.63 0-3.186-.317-4.666-.95-1.48-.634-2.756-1.486-3.827-2.557-1.07-1.071-1.923-2.347-2.557-3.827A11.736 11.736 0 0 1 2 14c0-1.63.317-3.186.95-4.666.634-1.48 1.486-2.756 2.557-3.827 1.071-1.07 2.347-1.923 3.827-2.557C10.814 2.317 12.369 2 14 2zm5.919 7c.298 0 .553.109.764.326.211.217.317.48.317.786 0 .307-.107.57-.32.79l-7.56 7.777a1.02 1.02 0 0 1-.76.321 1.03 1.03 0 0 1-.768-.321l-4.28-4.394a1.09 1.09 0 0 1-.312-.79c0-.307.106-.57.317-.787.21-.217.466-.325.764-.325s.555.11.769.33l3.51 3.611 6.79-6.994c.214-.22.47-.33.769-.33z"
    />
  </svg>
</button>
`;
